File: TasksList/Python/Web/Plain/src/storage/adapters/mysql.py
import os
import pymysql
from pymysql.cursors import DictCursor
from storage.database_adapter import DatabaseAdapter

TYPE_MAP = {
    "TEXT": "TEXT",
    "INTEGER": "INT",
    "TIMESTAMP": "DATETIME",
}

class MySQL(DatabaseAdapter):
    def __init__(self):
        self.conn = None
        self.connect()

    def connect(self):
        if self.conn and self.conn.open:
            return
        self.conn = pymysql.connect(
            host=os.getenv("DB_HOST", "db"),
            port=int(os.getenv("DB_PORT", "3306")),
            user=os.getenv("DB_USER", "root"),
            password=os.getenv("DB_PASSWORD", ""),
            database=os.getenv("DB_NAME", "tasks"),
        )

    def _translate(self, col_def):
        for generic, specific in TYPE_MAP.items():
            col_def = col_def.replace(generic, specific)
        return col_def

    def init_table(self, table, columns=None):
        self.connect()
        cur = self.conn.cursor()
        cols = ["id INT AUTO_INCREMENT PRIMARY KEY"]
        if columns:
            for name, col_def in columns.items():
                cols.append(f"{name} {self._translate(col_def)}")
        cur.execute(f"CREATE TABLE IF NOT EXISTS {table} ({', '.join(cols)})")
        self.conn.commit()
        cur.close()

    def _query(self, sql, params=None, fetchone=False, fetchall=False):
        self.connect()
        cur = self.conn.cursor(DictCursor)
        cur.execute(sql, params or ())
        if fetchone:
            row = cur.fetchone()
            cur.close()
            return dict(row) if row else None
        if fetchall:
            rows = [dict(r) for r in cur.fetchall()]
            cur.close()
            return rows
        cur.close()

    def _execute(self, sql, params=None):
        self.connect()
        cur = self.conn.cursor()
        cur.execute(sql, params or ())
        self.conn.commit()
        rowcount = cur.rowcount
        cur.close()
        return rowcount

    def get_all(self, table):
        return self._query(f"SELECT * FROM {table} ORDER BY id ASC", fetchall=True)

    def get_by_id(self, table, id):
        return self._query(f"SELECT * FROM {table} WHERE id = %s", (id,), fetchone=True)

    def create(self, table, data):
        cols = ", ".join(data.keys())
        placeholders = ", ".join("%s" for _ in data)
        values = list(data.values())
        self.connect()
        cur = self.conn.cursor(DictCursor)
        cur.execute(f"INSERT INTO {table} ({cols}) VALUES ({placeholders})", values)
        self.conn.commit()
        last_id = cur.lastrowid
        cur.execute(f"SELECT * FROM {table} WHERE id = %s", (last_id,))
        row = dict(cur.fetchone())
        cur.close()
        return row

    def update(self, table, id, data):
        set_clause = ", ".join(f"{k} = %s" for k in data)
        values = list(data.values()) + [id]
        return self._execute(f"UPDATE {table} SET {set_clause} WHERE id = %s", values) > 0

    def delete(self, table, id):
        return self._execute(f"DELETE FROM {table} WHERE id = %s", (id,)) > 0

    def search(self, table, field, query):
        return self._query(
            f"SELECT * FROM {table} WHERE LOWER({field}) LIKE %s ORDER BY id ASC",
            (f"%{query.lower()}%",), fetchall=True
        )

    def close(self):
        if self.conn and self.conn.open:
            self.conn.close()

File: PasswordGenerator/Node.js/Web/Plain/src/storage/adapters/MySQL.js
const DatabaseAdapter = require('../DatabaseAdapter');

class MySQLAdapter extends DatabaseAdapter {
  constructor(defaultDB) {
    super();
    this.defaultDB = defaultDB || 'mysql';
  }

  async connect() {
    const mysql = require('mysql2/promise');
    this.pool = mysql.createPool({
      host: process.env.DB_HOST || 'db',
      port: parseInt(process.env.DB_PORT || '3306'),
      database: process.env.DB_NAME || this.defaultDB,
      user: process.env.DB_USER || 'postgres',
      password: process.env.DB_PASSWORD || 'postgres',
    });
  }

  _colType(col) {
    switch (col.type) {
      case 'string': return 'VARCHAR(255)';
      case 'text': return 'TEXT';
      case 'integer': return 'INT';
      case 'boolean': return 'TINYINT';
      case 'datetime': return 'DATETIME';
      default: return 'VARCHAR(255)';
    }
  }

  _colDef(col) {
    let def = `${col.name} ${this._colType(col)}`;
    if (col.required) def += ' NOT NULL';
    if (col.default !== undefined) {
      if (col.default === 'now') def += ' DEFAULT CURRENT_TIMESTAMP';
      else if (typeof col.default === 'string') def += ` DEFAULT '${col.default}'`;
      else def += ` DEFAULT ${col.default}`;
    }
    return def;
  }

  async init(table, columns) {
    const cols = columns.map(c => this._colDef(c)).join(', ');
    await this.pool.execute(`CREATE TABLE IF NOT EXISTS ${table} (id INT AUTO_INCREMENT PRIMARY KEY, ${cols})`);
  }

  async getAll(table, options = {}) {
    let sql = `SELECT * FROM ${table}`;
    if (options.orderBy) sql += ` ORDER BY ${options.orderBy} ${options.orderDir || 'ASC'}`;
    if (options.limit) sql += ` LIMIT ${options.limit}`;
    const [r] = await this.pool.execute(sql);
    return r;
  }

  async getById(table, id) {
    const [r] = await this.pool.execute(`SELECT * FROM ${table} WHERE id = ?`, [id]);
    return r[0] || null;
  }

  async create(table, data) {
    const keys = Object.keys(data);
    const vals = Object.values(data);
    const placeholders = keys.map(() => '?').join(', ');
    const [r] = await this.pool.execute(
      `INSERT INTO ${table} (${keys.join(', ')}) VALUES (${placeholders})`,
      vals
    );
    const [rows] = await this.pool.execute(`SELECT * FROM ${table} WHERE id = ?`, [r.insertId]);
    return rows[0];
  }

  async update(table, id, data) {
    const keys = Object.keys(data);
    const vals = Object.values(data);
    const setClause = keys.map(k => `${k} = ?`).join(', ');
    vals.push(id);
    await this.pool.execute(`UPDATE ${table} SET ${setClause} WHERE id = ?`, vals);
    const [r] = await this.pool.execute(`SELECT * FROM ${table} WHERE id = ?`, [id]);
    return r[0] || null;
  }

  async delete(table, id) {
    const [r] = await this.pool.execute(`DELETE FROM ${table} WHERE id = ?`, [id]);
    return { changes: r.affectedRows };
  }

  async clear(table) {
    const [r] = await this.pool.execute(`DELETE FROM ${table}`);
    return { changes: r.affectedRows };
  }

  async search(table, field, query) {
    const [r] = await this.pool.execute(
      `SELECT * FROM ${table} WHERE LOWER(${field}) LIKE ? ORDER BY id ASC`,
      [`%${query.toLowerCase()}%`]
    );
    return r;
  }

  async close() {
    await this.pool.end();
  }
}
module.exports = MySQLAdapter;
